pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1312,3 +1378,136 @@ func updatePipelineRunStatusFromRuns(logger *zap.SugaredLogger, pr *v1beta1.Pipe
 		}
 	}
 }
+
+func updatePipelineRunStatusFromChildRefs(logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, trs []*v1beta1.TaskRun, runs []*v1alpha1.Run) map[string]string {
+	// If no TaskRun or Run was found, nothing to be done. We never remove child references from the status.
+	// We do still return an empty map of TaskRun/Run names keyed by PipelineTask name for later functions.
+	if len(trs) == 0 && len(runs) == 0 {
+		return make(map[string]string)
+	}
+
+	// Store a list of Condition TaskRuns for each PipelineTask (by name)
+	conditionTaskRuns := make(map[string][]*v1beta1.TaskRun)
+	// Map PipelineTask names to TaskRun names that were already in the status
+	taskRunNameByPipelineTask := make(map[string]string)
+	// Map PipelineTask names to TaskRun child references that were already in the status
+	childRefByPipelineTask := make(map[string]*v1beta1.ChildStatusReference)
+
+	for i := range pr.Status.ChildReferences {
+		childRefByPipelineTask[pr.Status.ChildReferences[i].PipelineTaskName] = &pr.Status.ChildReferences[i]
+		taskRunNameByPipelineTask[pr.Status.ChildReferences[i].PipelineTaskName] = pr.Status.ChildReferences[i].Name
+	}
+
+	// Loop over all the TaskRuns associated to Tasks
+	for _, tr := range trs {
+		// Only process TaskRuns that are owned by this PipelineRun.
+		// This skips TaskRuns that are indirectly created by the PipelineRun (e.g. by custom tasks).
+		if len(tr.OwnerReferences) < 1 || tr.OwnerReferences[0].UID != pr.ObjectMeta.UID {
```

Comment:
this block is repeated in a few places, can it be reused?

Reply:
It shows up twice for `TaskRun` and twice for `Run` - not sure if it's worth doing that as two functions, but I'm open to it.

Reply:
these two functions in general share a lot of code, could you experiment to see how it can be consolidated?

Reply:
Lemme see.

Reply:
So fwiw on this particular block, I just tried going with a helper function or two (i.e., either one function each for `TaskRun` and `Run`, or one generic one with a type `switch`) and it didn't really make things any clearer, since we still have to have a block here for the `continue`. I'm still seeing if there are any other shared bits I can consolidate, but this particular one doesn't feel worth doing.

Reply:
Actually, turns out that once I decided to try doing the filtering of `condition` `TaskRun`s before entering the main loop, I saw the obvious and realized that we could just filter out the "not owned by this `PipelineRun`" `TaskRun`s and `Run`s before entering the main loop as well. 🤦 =) How does https://github.com/tektoncd/pipeline/pull/4739/commits/6c8d61508a0aac2786cec9ae2e53e1114a91ade0 look?